### DISCUSSION TITLE
Serial ByteStreamDriverModel for replacing TcpClient comm uplink/downlink

### DISCUSSION THREAD

**Discussion by markospacetech** [nasa/fprime#1254] [Q&A]
Hi all,

I am currently working on a small sat project utilizing RockBlock radio components. I have written a component for the radio that parses and sends messages using read and write handlers over UART. Now instead of arbitrary messages I want to forward telemetry and commands through here. 

To achieve this I believe the best option is to have RockBlock implement the ByteStreamDriver Model similar to TcpClient. But I figured I'd come here for a second opinion / general thoughts (and for the chance that someone has an example that implements the ByteStreamDriver via serial driver rather than IP drivers).

My setup includes 2 raspberry pis, both with the same radio modules, but only one includes peripherals (Gps, camera, etc). My plan is to create a separate f' deployment for the RPI that is just handling the radio communication that receives/sends messages via radio with commands coming from the TcpClient. 

Here is an example of what I think the comm topologies should look like.

Rpi ground : 
```
                     tcpComm.Recv  ->  uplink.framedIn
                     uplink.comOut ->  RockBlock.sendPacket/buffer

                     RockBlock.receivePacket/buffer -> downlink.comIn
                     downlink.framedOut  ->  tcpComm.send
```
Rpi satellite: 
```
                      RockBlock.receivePacket/buffer  ->  uplink.framedIn  # deframe commands from ground control
                      uplink.comOut ->  cmdDisp.seqCmdBuff
                      
                      chanTlm.PktSend -> downlink.comIn                # frame telemetry values from components
                      downlink.framedOut -> RockBlock.sendPacket/buffer
```
                     
I think this is the best method but thoughts and recommendations here are more than welcome. Writing this out was also helpful for my understanding :sweat_smile: 

Cheers,
Marko :rocket:

**Comment by LeStarch** [comment]
This is what I'd do.  @timcanham thoughts?

**Reply by timcanham** [reply]
Yes - I think it should work. Just have to trace through the ports to make sure the data types all match.

**Reply by markospacetech** [reply]
Ok great. Makes sense in theory but a bit trickier than I thought :sweat_smile: . Would I need to implement the ByteStreamDriver as well as the LinuxSerialDriver or just the ByteStreamDriver and the Serial driver as a member object?

 In my component that I have already I connect the serialOut port from my uartDrv to the rockBlock serialRecv input (uartDrv.serialRecv -> rockBlock.serialRecv). Then as the rate group loops I store the data in a buffer until stop bytes are received from the radio to separate the messages. 

Not sure how to simulate this in my driver. I think I need to study the ByteStreamDriver a bit more first.

**Reply by timcanham** [reply]
When you say "stop bytes," do you mean serial bytes that are packet delimiters from the radio?

**Reply by timcanham** [reply]
Sort of "magic" values?

**Reply by markospacetech** [reply]
Every message from the radio ends with bytes '/n' '/r'. However when reading from the LinuxSerialDriver, I only received one byte at a time so I would append each byte to a buffer until the stop bytes are detected. Then I clear the buffer and start scanning for the next message.

**Reply by timcanham** [reply]
Do you need a rate group to do that? Can't you just accumulate with each port call from the serial driver and look for the characters? BTW, it's been my experience that you get data in chunks from the serial port rather than a byte at a time due to hardware buffers. Is your data rate very low such that it empties the buffers frequently?

**Reply by markospacetech** [reply]
Not sure if the rate group is necessary but I based it off of the RpiDemo. I believe I am accumulating with each port call though. How do I know what my data rate is?

Here is my SerialRead handler

```void RockBlockComponentImpl ::
    UartRead_handler(
        const NATIVE_INT_TYPE portNum,
        Fw::Buffer &serBuffer,
        Drv::SerialReadStatus &status
    )
  {
      // convert incoming data to string.
      char* bPointer = reinterpret_cast<char*>(serBuffer.getData());    

      if (serial_status != Drv::SerialReadStatus::SER_OK) {
          Fw::Logger::logMsg("[WARNING] Received buffer with bad packet: %d\n", serial_status.e);
          serBuffer.setSize(UART_READ_BUFF_SIZE);
          this->UartBuffers_out(0, serBuffer);
          return;
      }

      // data from serial typically outputs one byte at a time, build message until stop bytes detected
      for (NATIVE_UINT_TYPE byte = 0; byte < serBuffer.getSize(); byte++) {

                responseBuffer[rbBufferIndex] = bPointer[byte];

                // line feed and return carriage are stop bytes for separating messages
                if(rbBufferIndex > 1 && responseBuffer[rbBufferIndex] == '\n' && 
                                            responseBuffer[rbBufferIndex-1] == '\r'){                 
                    
                 
                  if(detectCommand("OK", responseBuffer)){
                     Fw::LogStringArg evrMsg(m_lastUartMsg);
                     this->log_ACTIVITY_HI_RD_UartMsgIn(evrMsg);
                    }
              
                    this->m_lastUartMsg = responseBuffer;
            

                    memset(responseBuffer, 0, RB_RESPONSE_BUFFER_SIZE);
                    rbBufferIndex = 0;

                    }else{
                     // increase message index   
                    rbBufferIndex++;
                }         
      }

      this->m_uartReadBytes += serBuffer.getSize();

      // reset buffer size
      serBuffer.setSize(RPI_UART_READ_BUFF_SIZE);
      // return buffer to driver
      this->UartBuffers_out(0, serBuffer);
  }
```

**Reply by markospacetech** [reply]
Sometimes it would receive more than one byte though, so I assumed it was a hardware thing

**Reply by timcanham** [reply]
Your `open()` call to the serial driver should tell you the rate. I think you have a race condition in your code. You shouldn't send the buffer back to the driver until you have pulled your data out of it, or it could get overwritten by new data from the UART. Move `this->UartBuffers_out(0, serBuffer);` to the end of the handler function.

**Reply by timcanham** [reply]
There is another possible bug:

You do:

```c++
serBuffer.setSize(UART_READ_BUFF_SIZE);
```

before:
```c++
for (NATIVE_UINT_TYPE byte = 0; byte < serBuffer.getSize(); byte++) {
```

The driver sets the buffer size to whatever data it gets before calling your handler. When you do a `setSize()` you lose how much data you actually got and set it to the full size. Your loop later will walk through the entire buffer, even if it got less data, and you could be processing junk data.

**Reply by markospacetech** [reply]
I do have 
```
      serBuffer.setSize(RPI_UART_READ_BUFF_SIZE);
      // return buffer to driver
      this->UartBuffers_out(0, serBuffer);
``` 
at the end 

and the first 
```
serBuffer.setSize(UART_READ_BUFF_SIZE);
```
only runs if there is a serial read error, which then returns before the rest of the function (from GpsApp example)

**Reply by timcanham** [reply]
Oops - see that now.

**Reply by markospacetech** [reply]
```
uartDrv.open("/dev/serial0",
            Drv::LinuxSerialDriverComponentImpl::BAUD_19200,
            Drv::LinuxSerialDriverComponentImpl::NO_FLOW,
            Drv::LinuxSerialDriverComponentImpl::PARITY_NONE,
            true))
```
Is the rate the Baud rate? This is the rate that the radio is sending on as well.

**Comment by timcanham** [comment]
Yes, your rate is 19200 bits/sec (aka baud). If your computer is relatively fast and the buffers are smaller, you could get it a byte at a time sometimes.

**Reply by markospacetech** [reply]
Ok that makes sense. So for the ByteStream implementation, I need to somehow build the message/buffer before deallocate_out(0, fwBuffer)? (looking at [TcpClientComponentImpl.cpp](https://github.com/nasa/fprime/blob/7e20febd34019c730da1358567e7a512592de4d8/Drv/TcpClient/TcpClientComponentImpl.cpp#L72) )

**Comment by ebenezerarunkumar** [comment]
@markospacetech were you able to complete a version of this serial driver that uses the ByteStreamDriverModel? I am looking to do the same thing and wanted to see if you had an open-source version I could reference!
